lib/sycamore/sycamore/utils/derivation_utils.py

```diff
@@ -0,0 +1,202 @@
+from collections import deque, defaultdict
+
+
+class UnitConverter:
+    def __init__(self):
+        # Graph representation: adjacency list with conversion rates
+        self.graph = defaultdict(dict)
+        self.add_conversion("SF", "SM", 0.092903)
+        self.add_conversion("acre", "SF", 43560.0)
+        self.add_conversion("hectare", "SM", 10000.0)
+        self.add_conversion("acre", "hectare", 0.404686)
+
+    def add_conversion(self, from_unit, to_unit, factor):
+        self.graph[from_unit][to_unit] = factor
+        self.graph[to_unit][from_unit] = 1 / factor  # Inverse
+
+    def find_conversion_rate(self, from_unit, to_unit):
+        # BFS to find shortest path and cumulative conversion factor
+        if from_unit == to_unit:
+            return 1.0
+        visited = set()
+        queue = deque([(from_unit, 1.0)])  # (current_unit, cumulative_factor)
+        while queue:
+            current_unit, current_factor = queue.popleft()
+            visited.add(current_unit)
+            for neighbor, factor in self.graph[current_unit].items():
+                if neighbor == to_unit:
+                    return current_factor * factor
+                if neighbor not in visited:
+                    queue.append((neighbor, current_factor * factor))
+        raise ValueError(f"No conversion path found between {from_unit} and {to_unit}")
+
+    def convert(self, value, from_unit, to_unit):
+        rate = self.find_conversion_rate(from_unit, to_unit)
+        return round(value * rate, 6)
+
+    def check(self, from_unit, to_unit, value1, value2):
+        converted_value = self.convert(value1, from_unit, to_unit)
+        return value2 and abs(converted_value - value2) / value2 < 0.05
+
+
+class UnitDerivation:
```

Comment:
On further inspection, I don't think this class is necessary, see comment below in PropertyDerivation.